File: solo-leveling-player-service/src/main/java/com/sleepkqq/sololeveling/player/service/repository/user/UserRepository.java
package com.sleepkqq.sololeveling.player.service.repository.user;

import static com.sleepkqq.sololeveling.player.service.model.Tables.USER_TABLE;

import com.sleepkqq.sololeveling.player.service.model.user.User;
import java.util.Optional;
import org.babyfish.jimmer.spring.repository.JRepository;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Repository
public interface UserRepository extends JRepository<User, Long> {

  @Transactional
  default Optional<Integer> findVersionById(long id) {
    var table = USER_TABLE;
    return sql().createQuery(table)
        .where(table.id().eq(id))
        .select(table.version())
        .fetchOptional();
  }
}

File: solo-leveling-player-model/src/main/kotlin/com/sleepkqq/sololeveling/player/model/repository/user/UserRepository.kt
package com.sleepkqq.sololeveling.player.model.repository.user

import com.sleepkqq.sololeveling.player.model.entity.user.User
import com.sleepkqq.sololeveling.player.model.entity.user.id
import com.sleepkqq.sololeveling.player.model.entity.user.version
import org.babyfish.jimmer.spring.repository.KRepository
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.springframework.stereotype.Repository

@Repository
interface UserRepository : KRepository<User, Long> {

	fun findVersionById(id: Long): Int? =
		sql.createQuery(User::class) {
			where(table.id eq id)
			select(table.version)
		}
			.fetchFirstOrNull()
}
